distributed/protocol/cuda.py

```diff
@@ -4,35 +4,36 @@
 import cupy
 from .serialize import dask_serialize, dask_deserialize
 
-# Some questions
-# 1.Do we need *protocol-dependent* serialization?
-#   I assume we want this kind of serialization only when
-#   in UCP.
-# 2. What does ucp-py need to know about?
-
 
 @dask_serialize.register(cupy.ndarray)
 def serialize_cupy_ndarray(x):
     # TODO: handle non-contiguous
-    # shape
-    # typestr
-    # descr
-    # data
-    # version
-    # strides (noncontiguous-only)
+    # TODO: Handle order='K' ravel
+    # TODO: 0d
+
+    if x.flags.c_contiguous or x.flags.f_contiguous:
+        strides = x.strides
+        data = x.ravel()  # order='K'
+    else:
+        x = cupy.ascontiguousarray(x)
+        strides = x.strides
+        data = x.ravel()
+
+    dtype = (0, x.dtype.str)
+
     header = x.__cuda_array_interface__.copy()
-    header['device'] = x.device.id
     header['lengths'] = (x.nbytes,)  # one per stride
     header['compression'] = (None,)  # TODO
-    # TODO: I don't think ucx-py should have to worry about
-    # MemoryPointer. Maybe some thin wrapper.
-    return header, [x.data]
+    header['is_cuda'] = 1
```

Comment:
The approach suggested above would also stop the various serialization routines from having to know about device vs host frames, which will probably help future people who have to implement these.